### PR TITLE
planner: add some test cases about partition table dynamic-mode and plan-cache

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: related to #24150 

Problem Summary: planner: add some test cases about partition table dynamic-mode and plan-cache

### What is changed and how it works?

No logical change, just add some test cases to test that plan-cache can work when enabling partition table dynamic-mode.

### Check List 

Tests 

- Unit test

### Release note 

- planner: add some test cases about partition table dynamic-mode and plan-cache

### CODE REVIEW

File: planner/core/prepare_test.go [hunk 1]
```diff
@@ -1193,3 +1196,130 @@ func (s *testPlanSerialSuite) TestIssue23671(c *C) {
...
+	partition p3 values less than ('800'),
+	partition p4 values less than ('9999'))`,
+			`create table t2(a varchar(10), b varchar(10))`,
+			func() string { return fmt.Sprintf(`("%v", "%v")`, rand.Intn(1000), rand.Intn(1000)) },
+			func() string { return fmt.Sprintf(`"%v"`, rand.Intn(1000)) },
+			`select * from %v where a > ?`,
+		},
+		{ // range partition + datetime
+			`create table t1(a datetime, b datetime) partition by range columns(a) (
+	partition p0 values less than ('1970-1-1 00:00:00'),
```
Comment (mjonss) [inline planner/core/prepare_test.go:1273]: Please use [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format with prefix 0 for dates, like 1970-01-01.
Also good to test the 'zero' epoch 🌟

**Review by mjonss — COMMENTED** [review]
Should also timestamp be tested?
Otherwise LGTM (but I'm not yet promoted to reviewer).

File: planner/core/prepare_test.go (hunk 1 above)
Comment (qw4990) [reply planner/core/prepare_test.go:1273]: OK, I'll change it and add more cases about `timestamp` later.

**Comment by qw4990** [comment]
> Should also timestamp be tested?
> Otherwise LGTM (but I'm not yet promoted to reviewer).

I find that `timestamp` is not supported by range partition, so I ignore this case, but I add one more case for `date`.
Other comments are addressed.
And also PTAL @rebelice @time-and-fate

**Review by rebelice — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]
/lgtm

**Pull request merged**